fmp_server.py

```diff
@@ -305,18 +305,51 @@ def recvReqLayerUp(self, packet_id, data, seq):
          - Relocate > Select Server
         """
         if self.session.layer == 2:
-            city = self.session.get_city()
-            if city.leader == self.session:
-                if len(city.players) > 1:
-                    # TODO: Transfer the leadership to another player
-                    pass
-                city.leader = None
+            new_leader = self.session.try_transfer_city_leadership()
+            if new_leader is not None:
+                self.sendNtcLayerHost(new_leader, seq)
 
             ntc_data = pati.lp2_string(self.session.capcom_id)
             self.server.layer_broadcast(self.session, PatID4.NtcLayerOut,
                                         ntc_data, seq)
         self.sendAnsLayerUp(data, seq)
 
+    def recvReqLayerEnd(self, packet_id, data, seq):
```

Comment:
You're shadowing a method from the `pat` module, if you're sure that's only used by the FMP server and not the other ones, you should remove the implementation in the `pat` module. Otherwise, I'd suggest you move your code in the `pat` module instead.

Reply:
Fixed in latest commit.